File: clients/python/codegame/client_message.py
from debugging.debug_command import DebugCommand
from model.action import Action
from stream_wrapper import StreamWrapper

class ClientMessage:
    """Message sent from client"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "ClientMessage":
        """Read ClientMessage from input stream
        """
        tag = stream.read_int()
        if tag == DebugMessage.TAG:
            return ClientMessage.DebugMessage.read_from(stream)
        if tag == ActionMessage.TAG:
            return ClientMessage.ActionMessage.read_from(stream)
        if tag == DebugUpdateDone.TAG:
            return ClientMessage.DebugUpdateDone.read_from(stream)
        if tag == RequestDebugState.TAG:
            return ClientMessage.RequestDebugState.read_from(stream)
        raise Exception("Unexpected tag value")

class DebugMessage(ClientMessage):
    """Ask app to perform new debug command"""

    TAG = 0

    __slots__ = ("command",)

    command: DebugCommand

    def __init__(self, command: DebugCommand):
        self.command = command
        """Command to perform"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "DebugMessage":
        """Read DebugMessage from input stream
        """
        command = DebugCommand.read_from(stream)
        return DebugMessage(command)
    
    def write_to(self, stream: StreamWrapper):
        """Write DebugMessage to output stream
        """
        stream.write_int(self.TAG)
        self.command.write_to(stream)
    
    def __repr__(self):
        return "DebugMessage(" + \
            repr(self.command) + \
            ")"

ClientMessage.DebugMessage = DebugMessage

class ActionMessage(ClientMessage):
    """Reply for ServerMessage::GetAction"""

    TAG = 1

    __slots__ = ("action",)

    action: Action

    def __init__(self, action: Action):
        self.action = action
        """Player's action"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "ActionMessage":
        """Read ActionMessage from input stream
        """
        action = Action.read_from(stream)
        return ActionMessage(action)
    
    def write_to(self, stream: StreamWrapper):
        """Write ActionMessage to output stream
        """
        stream.write_int(self.TAG)
        self.action.write_to(stream)
    
    def __repr__(self):
        return "ActionMessage(" + \
            repr(self.action) + \
            ")"

ClientMessage.ActionMessage = ActionMessage

class DebugUpdateDone(ClientMessage):
    """Signifies finish of the debug update"""

    TAG = 2

    __slots__ = ()


    def __init__(self):
        pass

    @staticmethod
    def read_from(stream: StreamWrapper) -> "DebugUpdateDone":
        """Read DebugUpdateDone from input stream
        """
        return DebugUpdateDone()
    
    def write_to(self, stream: StreamWrapper):
        """Write DebugUpdateDone to output stream
        """
        stream.write_int(self.TAG)
    
    def __repr__(self):
        return "DebugUpdateDone(" + \
            ")"

ClientMessage.DebugUpdateDone = DebugUpdateDone

class RequestDebugState(ClientMessage):
    """Request debug state from the app"""

    TAG = 3

    __slots__ = ()


    def __init__(self):
        pass

    @staticmethod
    def read_from(stream: StreamWrapper) -> "RequestDebugState":
        """Read RequestDebugState from input stream
        """
        return RequestDebugState()
    
    def write_to(self, stream: StreamWrapper):
        """Write RequestDebugState to output stream
        """
        stream.write_int(self.TAG)
    
    def __repr__(self):
        return "RequestDebugState(" + \
            ")"

ClientMessage.RequestDebugState = RequestDebugState

File: clients/rust/src/codegame/client_message.rs
use super::*;

/// Message sent from client
#[derive(Clone, Debug)]
pub enum ClientMessage {
    /// Ask app to perform new debug command
    DebugMessage {
        /// Command to perform
        command: debugging::DebugCommand,
    },
    /// Reply for ServerMessage::GetAction
    ActionMessage {
        /// Player's action
        action: model::Action,
    },
    /// Signifies finish of the debug update
    DebugUpdateDone {
    },
    /// Request debug state from the app
    RequestDebugState {
    },
}

impl trans::Trans for ClientMessage {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        match self {
            Self::DebugMessage {
                command,
            } => {
                <i32 as trans::Trans>::write_to(&0, writer)?;
                command.write_to(writer)?;
            }
            Self::ActionMessage {
                action,
            } => {
                <i32 as trans::Trans>::write_to(&1, writer)?;
                action.write_to(writer)?;
            }
            Self::DebugUpdateDone {
            } => {
                <i32 as trans::Trans>::write_to(&2, writer)?;
            }
            Self::RequestDebugState {
            } => {
                <i32 as trans::Trans>::write_to(&3, writer)?;
            }
        }
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let tag = <i32 as trans::Trans>::read_from(reader)?;
        match tag {
            0 => {
                let command: debugging::DebugCommand = trans::Trans::read_from(reader)?;
                Ok(Self::DebugMessage {
                    command,
                })
            }
            1 => {
                let action: model::Action = trans::Trans::read_from(reader)?;
                Ok(Self::ActionMessage {
                    action,
                })
            }
            2 => {
                Ok(Self::DebugUpdateDone {
                })
            }
            3 => {
                Ok(Self::RequestDebugState {
                })
            }
            _ => Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Unexpected tag {:?}", tag))),
        }
    }
}